packages/hash/api/src/graphql/resolvers/pages/updatePageContents.ts

```diff
@@ -1,70 +1,239 @@
 import { ApolloError, UserInputError } from "apollo-server-errors";
-
+import { produce } from "immer";
 import {
-  Resolver,
+  Block,
+  Entity,
+  EntityType,
+  Page,
+  UnresolvedGQLEntity,
+} from "../../../model";
+import { exactlyOne } from "../../../util";
+import {
+  EntityDefinition,
   MutationUpdatePageContentsArgs,
-  UpdatePageAction,
-  UpdateEntity,
+  Resolver,
   SwapBlockData,
+  UpdateEntity,
+  UpdatePageAction,
+  UpdatePageContentsResult,
 } from "../../apiTypes.gen";
-import { Block, Entity, Page, UnresolvedGQLEntity } from "../../../model";
 import { LoggedInGraphQLContext } from "../../context";
-import { exactlyOne } from "../../../util";
 
 const validateActionsInput = (actions: UpdatePageAction[]) => {
   for (const [i, action] of actions.entries()) {
     if (
       !exactlyOne(
-        action.insertNewBlock,
+        action.insertBlock,
         action.moveBlock,
         action.removeBlock,
         action.updateEntity,
         action.swapBlockData,
+        action.createEntity,
+        action.createEntityType,
       )
     ) {
       throw new UserInputError(
-        `at action ${i}: exactly one of insertNewBlock, moveBlock, removeBlock or updateEntity must be specified`,
+        `at action ${i}: exactly one of insertBlock, moveBlock, removeBlock or updateEntity must be specified`,
       );
     }
   }
 };
 
+const isPlaceholder = (id: unknown): id is string =>
```

Comment:
Nit:
```suggestion
const isPlaceholderId = (id: unknown): id is string =>
```
I think we should call this `isPlaceholderId` and generally use `placeholderId` instead of `placeholder`. That way it's a bit consistent with how we name other ids => `draftId`, `entityId`.  At first glance I thought `placeholder` was some kind of draft entity. It's easier to infer it's a string if it's named `placeholderId`

Reply:
👍